### PR TITLE
Combinatorics: Added naive algorithm for computing conjugacy classes in permutation groups

### PR DESCRIPTION
**Pull request by sachin-4099**
Reference: https://github.com/sympy/sympy/pull/9160

#### Brief description of what is fixed or changed
This pull request adds `conjugacy_class` and `conjugacy_classes` methods to the `PermutationGroup class`, which compute respectively the set of all elements conjugate to an element of a group, and the list of all these equivalence classes.

#### Other comments
Regression tests are also added

#### Release Notes

* combinatorics
  * Added naive algorithm for computing conjugacy classes in permutation groups

### CODE REVIEW

**Comment by sachin-4099** [comment]
I have just revived this PR: https://github.com/sympy/sympy/pull/9160

**Comment by sachin-4099** [comment]
I just want to know if adding conjugacy classes would be a good idea or is this already implemented somewhere else. @oscarbenjamin @certik

**Comment by sachin-4099** [comment]
According to @lucjon, https://github.com/sympy/sympy/pull/9160#issue-31267916:

This implementation finds the conjugacy classes in a relatively naïve way: by directly computing the orbit of each element under conjugation. 1, written in 2006, states that this is efficient for groups up to order 10^4. The naïve algorithm is also very general, in that it doesn't rely on any properties of the underlying group representation.

There are more complex algorithms which make equivalent computation more feasible (see 2, or 3 for an unredacted version behind a paywall) for larger groups. These would require in particular that conjugacy_classes did not simply return a list of all the elements in each conjugacy class; however, it is debatable whether this is even a worthwhile representation for the naïve implementation. Should the patch be amended to have it return only, say, a list of representatives?

**Comment by sachin-4099** [comment]
Here is a bit faster algorithm to compute the same by @pernici :
https://github.com/sympy/sympy/pull/9160#issuecomment-82567054

**Comment by sylee957** [comment]
I think that it would be useful to have features like conjugacy classes, or finding all subgroups, even if it is computationally naive or expensive at some stage.

But you should better rebase from the original PR than copy pasting. And I would expect the docstrings to be spaced like the original.

**Comment by sachin-4099** [comment]
> And I would expect the docstrings to be spaced like the original.
What do you mean by this? What changes do I have to make?

File: sympy/combinatorics/perm_groups.py [hunk 1]
```diff
@@ -2567,6 +2567,78 @@ def minimal_block(self, points):
...
+        Note that conjugacy is an equivalence relation, and therefore that
+        conjugacy classes are partitions of ``G``. For a list of all the
+        conjugacy classes of the group, use the conjugacy_classes() method.
+        In a permutation group, each conjugacy class corresponds to a particular
+        `cycle structure': for example, in ``S_3``, the conjugacy classes are:
+            * the identity class, ``{()}``
+            * all transpositions, ``{(1 2), (1 3), (2 3)}``
+            * all 3-cycles, ``{(1 2 3), (1 3 2)}``
+        Examples
+        ========
```
Comment (sylee957) [inline sympy/combinatorics/perm_groups.py]: https://github.com/sympy/sympy/pull/18350#issuecomment-575004951
I mean in the original one, it was spaced like 
```
 * all 3-cycles, ``{(1 2 3), (1 3 2)}``

Examples
========

>>> from sympy.combinatorics.permutations import Permutation
```
And now, you have made them less readable.

**Comment by sachin-4099** [comment]
> I think that it would be useful to have features like conjugacy classes, or finding all subgroups, even if it is not the best.
> 
> But you should better rebase from the original PR than copy pasting. And I would expect the docstrings to be spaced like the original.

I tried to rebase, but i think some error occured. Can you please explain the exact procedure?

**Comment by sylee957** [comment]
You should 
1. Add @lucjon as a remote and checkout `lucjon/conj-grp`.
2. From his branch, create a new branch locally via `git checkout -b` and use `git rebase -i master`
3. Resolve merge conflicts.

**Comment by sachin-4099** [comment]
> You should
> 
> 1. Add @lucjon as a remote and checkout `lucjon/conj-grp`.
> 2. From his branch, create a new branch locally via `git checkout -b` and use `git rebase -i master`
> 3. Resolve merge conflicts.

Please tell the exact commands in the order of execution, otherwise again some issue will occur.

**Comment by sylee957** [comment]
Okay, 
1. `git remote add lucjon https://github.com/lucjon/sympy.git`
2. `git fetch lucjon`
3. `git checkout lucjon/conj-grp`
4. `git checkout -b conj-grp`
5. `git rebase -i master`

**Comment by sachin-4099** [comment]
pick 815317c99a Add naive algorithm for computing conjugacy classes in perm groups
pick cba94ff98a Fix incorrect example/doctest in PermutationGroup#conjugacy_class
pick 6673af6183 Fix whitespace issues.

 Rebase bbd670d781..6673af6183 onto bbd670d781 (3 commands)

 Commands:
 p, pick <commit> = use commit
 r, reword <commit> = use commit, but edit the commit message
 e, edit <commit> = use commit, but stop for amending
 s, squash <commit> = use commit, but meld into previous commit
 f, fixup <commit> = like "squash", but discard this commit's log message
 x, exec <command> = run command (the rest of the line) using shell
 b, break = stop here (continue rebase later with 'git rebase --continue')
 d, drop <commit> = remove commit
 l, label <label> = label current HEAD with a name
 t, reset <label> = reset HEAD to a label
 m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
 .       create a merge commit using the original merge commit's
 .       message (or the oneline, if no original merge commit was
 .       specified). Use -c <commit> to reword the commit message.

 These lines can be re-ordered; they are executed from top to bottom.

 If you remove a line here THAT COMMIT WILL BE LOST.

 However, if you remove everything, the rebase will be aborted.

 Note that empty commits are commented out

**Comment by sachin-4099** [comment]
I should just save and quit.

**Comment by sachin-4099** [comment]
All the merge conflicts are resolved ans it says:
Successfully rebased and updated refs/heads/conj-grp.

**Comment by sachin-4099** [comment]
Now shall i push this commit to this branch?

**Comment by sachin-4099** [comment]
It states Everything up to date, when i try to force-push to this branch.

**Comment by sylee957** [comment]
I guess something had gone wrong as this branch contains no commits. 

But if you have the successful rebase locally, you should be able to open up another PR

**Comment by sachin-4099** [comment]
What should I do? Git is not able to identify the new changes. It states working tree clean.

**Comment by sylee957** [comment]
Can you see commits like `Add naive algorithm for computing conjugacy classes in perm groups` when you check `git log`?

**Comment by sachin-4099** [comment]
> Can you see commits like `Add naive algorithm for computing conjugacy classes in perm groups` when you check `git log`?

No issue I will open a new PR.

**Comment by sachin-4099** [comment]
I have opened a new PR: https://github.com/sympy/sympy/pull/18354

**Pull request closed without merging**